[SymmMem] Add all-to-all

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #151993
* #151819
* __->__ #151498
* #151261

Add an all-to-all impl based on NVSHMEM's on-stream API `nvshmemx_alltoallmem_on_stream`.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

kwen2501 committed:
[WIP][SymmMem] Add all-to-all

[ghstack-poisoned]

Skylion007 commented on torch/csrc/distributed/c10d/nvshmem_extension.cu:
```diff
@@ -452,4 +450,24 @@ at::Tensor nvshmem_sendrecv(
+at::Tensor nvshmem_all_to_all(
+    at::Tensor& input,
+    at::Tensor& out,
+    std::string group_name) {
```
Does group_name need to be a string by value or can it be a const ref or stringview?

kwen2501 replied:
Yeah I tried, but dispatcher signature registration does not seem to support `const std::string&`

kwen2501 committed:
Update on "[WIP][SymmMem] Add all-to-all"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 committed:
Update on "[WIP][SymmMem] Add all-to-all"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 committed:
Update on "[WIP][SymmMem] Add all-to-all"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 committed:
Update on "[WIP][SymmMem] Add all-to-all"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 committed:
Update on "[WIP][SymmMem] Add all-to-all"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 committed:
Update on "[WIP][SymmMem] Add all-to-all"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 committed:
Update on "[WIP][SymmMem] Add all-to-all"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 committed:
Update on "[WIP][SymmMem] Add all-to-all"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj reviewed (commented):
Shall we add a unit test for this?

kwen2501 committed:
Update on "[SymmMem] Add all-to-all"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 commented:
@fduwjj Yeah, can you give it a try?

kwen2501 committed:
Update on "[SymmMem] Add all-to-all"


Add an all-to-all impl based on NVSHMEM's on-stream API `nvshmemx_alltoallmem_on_stream`.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented on test/distributed/test_nvshmem.py:
```diff
@@ -0,0 +1,107 @@
+
+    @skipIfRocm
+    def test_nvshmem_all_to_all(self) -> None:
+        group_name = dist.group.WORLD.group_name
+        symm_mem.enable_symm_mem_for_group(group_name)
```
this will be called when you call rendezvous, no?

kwen2501 replied:
I guess so.

fduwjj commented on torch/csrc/distributed/c10d/SymmetricMemory.cpp:
```diff
@@ -276,6 +276,8 @@ TORCH_LIBRARY_FRAGMENT(symm_mem, m) {
       "memset32_(Tensor(a!) input, int offset, int val, int count) -> Tensor(a!)");
 
   m.def("nvshmem_broadcast(Tensor(a!) input, str group_name) -> Tensor(a!)");
+  m.def(
```
For my learning, this is pybind registration right?

kwen2501 replied:
Yeah, similar. A method would be added to `torch.ops.symm_mem`.

fduwjj reviewed (approved):
LGTM

kwen2501 committed:
Update on "[SymmMem] Add all-to-all"


Add an all-to-all impl based on NVSHMEM's on-stream API `nvshmemx_alltoallmem_on_stream`.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge
